consensus/tendermint/tmdb.go

```diff
@@ -0,0 +1,394 @@
+package tendermint
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	db "github.com/NethermindEth/juno/db"
+	"github.com/cockroachdb/pebble/v2"
+	"github.com/fxamacker/cbor/v2"
+)
+
+// Example DB layout for storing Tendermint WAL messages per height:
+//
+// We use two primary key prefixes based on db.BucketConsensus:
+//
+// 1. db.WALEntryCount: Stores the total number of messages (proposals, votes, timeouts)
+//    for a given height.
+//    - Key: db.WALEntryCount + <heightBytes>
+//    - Value: <countBytes>
+//    - Acts as a 1-based index for the individual message keys.
+//
+// 2. db.WALEntry: Stores the actual CBOR-encoded wrapped messages or timeouts.
+//    - Key: db.WALEntry + <heightBytes> + <iterBytes>
+//    - Value: CBOR-encoded wrapped Msg/Timeout
+//    - <iterBytes> ranges from 1 to the count stored under db.WALEntryCount.
+//
+// Both <heightBytes> and <iterBytes> are 4-byte big-endian uint32 representations.
+//
+// Batching:
+// We use a Batch (db.IndexedBatch) to accumulate writes before committing them to the DB.
+// This reduces expensive disk I/O. Reads check the batch first.
+// WARNING: If the process crashes before Commit(), buffered messages are lost,
+// risking consensus issues like equivocation (missing votes, double-signing).
+
+const NumBytesForHeight = 4
+
+type wrappedMsg struct {
+	Type string          `cbor:"type"`
+	Data cbor.RawMessage `cbor:"data"` // cbor serialised Msg or Timeout
+}
+
+// MessageType represents the type of message stored in the WAL.
+type MessageType uint8
+
+const (
+	MessageTypeProposal MessageType = iota
+	MessageTypePrevote
+	MessageTypePrecommit
+	MessageTypeTimeout
+	MessageTypeUnknown
+)
+
+// String returns the string representation of the MessageType.
+func (m MessageType) String() string {
+	switch m {
+	case MessageTypeProposal:
+		return "proposal"
+	case MessageTypePrevote:
+		return "prevote"
+	case MessageTypePrecommit:
+		return "precommit"
+	case MessageTypeTimeout:
+		return "timeout"
+	default:
+		return "unknown"
+	}
+}
+
+// TMDBInterface defines the methods for interacting with the Tendermint WAL database.
+type TMDBInterface[V Hashable[H], H Hash, A Addr] interface {
+	// CommitBatch writes the accumulated batch operations to the underlying database.
+	CommitBatch() error
+	// GetWALCount retrieves the number of WAL messages stored for a given height.
+	GetWALCount(height height) (uint32, error)
+	// GetWALMsgs retrieves all WAL messages (consensus messages and timeouts) stored for a given height from the database.
+	GetWALMsgs(height height) ([]IsWALMsg, error)
+	// SetWALEntry schedules the storage of a WAL message in the batch.
+	SetWALEntry(entry IsWALMsg, height height) error
+	// DeleteWALMsgs schedules the deletion of all WAL messages for a specific height in the batch.
+	DeleteWALMsgs(height height) error
+}
+
+// TMDB provides database access for Tendermint consensus state.
+type TMDB[V Hashable[H], H Hash, A Addr] struct {
```

Comment:
I think TM isn't a good abbreviation for Tendermint
```suggestion
type TendermintDB[V Hashable[H], H Hash, A Addr] struct {
```

Reply:
Updated